mcf-browser/src/back-end/parse.ts

```diff
@@ -0,0 +1,239 @@
+        ]);
+      }
+
+      finalReturn['localNodes'] = finalReturn['localNodes'].concat(
+          mcfOut['localNodes'],
+      );
+    } else if (fileExt === 'csv') {
+      if (tmcfFile) {
+        const tmcfFileName = (tmcfFile as File).name;
+        const tmcfOut = await ParseTmcf.generateMcf(tmcfFile, file).then(
+            (mcf) => {
+              const mcfParser = new ParseMcf(
+                  tmcfFileName + '&' + fileName,
+              );
+              return mcfParser.parseMcfStr(mcf as string);
+            },
+        );
+
+        const datapoints = await ParseTmcf.generateDataPoints(tmcfFile, file);
+        finalReturn['datapoints'] = mergeDataPoints(
+            finalReturn['datapoints'],
+            datapoints,
+        );
+
+        if (tmcfOut['errMsgs'].length !== 0) {
+          finalReturn['errMsgs'] = finalReturn['errMsgs'].concat({
+            file: tmcfFileName,
+            errs: tmcfOut['errMsgs'],
+          });
+        }
+        finalReturn['localNodes'] = finalReturn['localNodes'].concat(
+            tmcfOut['localNodes'],
+        );
+      }
+    }
+  }
+  return finalReturn;
+}
+
+/** Group nodes and find all time series
+  * @param {TimeDataObject} datapoints the time series data
+  * @return {Series[]} an array of time series in the data
+  */
+function getTimeData(datapoints: TimeDataObject) : TimeDataOutput {
+  // Turn from object to a list of series
+  const output: TimeDataOutput = {
+    timeData: [],
+    errMsgs: [],
+  };
+
+  const allSeries = Object.keys(datapoints);
+  for (const series of allSeries) {
+    const seriesObject = datapoints[series];
+    if (seriesObject) {
+      const parsedSeriesObject = parseSeries(series, seriesObject);
+      if (parsedSeriesObject.errMsgs.length > 0) {
+        // If there was an error, add it
+        output.errMsgs =
+            output.errMsgs.concat(parsedSeriesObject.errMsgs);
+      } else if (parsedSeriesObject.series) {
+        // If there is a successfully created Series object, add it
+        output.timeData.push(parsedSeriesObject.series);
+      }
+    }
+  }
+
+  return output;
+}
+
+/**
+  * Takes in the facet string generated when parsing the file
+  * and return an object of type Series
+  * @param {string} facet the facet defining the series
+  * @param {SeriesObject} values the values for the series
+  * @return {TimeDataOutput} a series object or an error message
+  */
+function parseSeries(facet: string, values: SeriesObject) : ParseSeriesOutput {
+  const {
+    variableMeasured,
+    observationAbout,
+    provenance,
+    measurementMethod,
+    observationPeriod,
+    unit,
+    scalingFactor,
+  } = Series.fromID(facet);
+  const data = [];
+  for (const date of Object.keys(values)) {
+    data.push({
+      x: date,
+      y: parseFloat((values as any)[date]),
+    });
+  }
+
+  if (variableMeasured === '' || observationAbout === '') {
+    return {
+      errMsgs: [],
+      series: null,
+    };
+  } else {
```

Comment:
nit: if you're returning in the `if` condition, could reduce indent for remaining code by dropping the `else`

Reply:
it's changed!